### PR TITLE
Allow subnamespace deletion if parent/anchor is missing

### LINKED ISSUE #847: HNC: allow direct deletion of a subnamespace if its parent doesn't exist
/assign @yiqigao217 

This comes up in some of the e2e tests - a namespace with the subnamespaceOf annotation exists, but its parent doesn't. We should allow it to be deleted.

### PR DESCRIPTION
**Pull request by yiqigao217**
Since we allow changes to the subnamespaceOf annotation on subnamespace
as an escape hatch and there's a scenario of recovering from CRD
deletion, this commit makes it easier to delete a subnamespace, but only
when the parent or the anchor is missing. If the anchor is ok (state),
we still don't allow direct deletion of a subnamespace and only allows
deletion from the anchor.

Tested with new unit test cases and manually before and after.

Fixes #847

### CODE REVIEW

File: incubator/hnc/internal/validators/namespace.go [hunk 1]
```diff
@@ -64,19 +64,15 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
 // handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
 // tested (ie without constructing a full admission.Request).
 func (v *Namespace) handle(req *nsRequest) admission.Response {
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
 	parent := req.ns.Annotations[api.SubnamespaceOf]
-	if parent != "" && req.op == v1beta1.Delete {
+	if parent != "" && v.Forest.Get(parent).HasAnchor(req.ns.Name) && req.op == v1beta1.Delete {
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/namespace.go:71]: This needs a lot more commenting IMO plus maybe a bit of refactoring. E.g. something like:

```
// If the anchor doesn't exist, we want to allow it to be deleted anyway because ...
// See issue #xxx
anchorExists := v.Forest.Get(parent).HasAnchor(req.ns.Name) <-- should be false if this isn't a subns, I think
if req.op == v1beta1.Delete && anchorExists {
   msg := ...
   return deny(forbidden, msg)
}
```

File: incubator/hnc/internal/validators/namespace.go [hunk 2]
```diff
@@ -64,19 +64,15 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
...
 		return deny(metav1.StatusReasonForbidden, msg)
 	}
 
-	return v.checkForest(req)
-}
-
-func (v *Namespace) checkForest(req *nsRequest) admission.Response {
-	v.Forest.Lock()
-	defer v.Forest.Unlock()
-
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/namespace.go:79]: This function is already much too long IMO, I don't think we should be making it longer. What's the motivation behind merging them?

If anything, maybe hold off on this change, submit another PR to restructure this webhook so each function is much smaller (maybe max 40 lines) and then reapply this change?

Comment (yiqigao217) [reply incubator/hnc/internal/validators/namespace.go:79]: Ah the only reason I did so is because we need to check (lock) the forest for that part too now before `checkForest()` func.
sgtm, I will work on the restructure first then come back to this PR.

File: incubator/hnc/internal/validators/namespace.go (hunk 1 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/namespace.go]: sg. updated!

File: incubator/hnc/internal/validators/namespace.go (hunk 2 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/namespace.go]: Done in #870

File: incubator/hnc/internal/validators/namespace_test.go [hunk 3]
```diff
@@ -12,29 +12,53 @@ import (
 )
 
 func TestDeleteSubNamespace(t *testing.T) {
-	// Create a namespace with owner annotation.
+	// Create a namespace instance a.
 	sub := &corev1.Namespace{}
 	sub.Name = "a"
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/namespace_test.go:17]: Maybe move this into the `t.Run` portion of the test? That's where we actually use it, plus it will ensure we get a fresh copy for each subtest.

File: incubator/hnc/internal/validators/namespace_test.go [hunk 4]
```diff
@@ -12,29 +12,53 @@ import (
...
-	f := foresttest.Create("B-")
-	vns := &Namespace{Forest: f}
+	// We will put an anchor a in the namespace c if it exists.
+	tests := []struct {
+		name           string
+		forest         string
+		subnamespaceOf string
+		fail           bool
+	}{
+		// There's no test case for when parent and anchor don't match, because the
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/namespace_test.go:26]: Do you mean: "when parent and annotation don't match"?

File: incubator/hnc/internal/validators/namespace_test.go [hunk 5]
```diff
@@ -12,29 +12,53 @@ import (
...
-	setSubAnnotation(sub, "b")
 
-	// a (subnamespace of b) -> b
-	f := foresttest.Create("B-")
-	vns := &Namespace{Forest: f}
+	// We will put an anchor a in the namespace c if it exists.
+	tests := []struct {
+		name           string
+		forest         string
+		subnamespaceOf string
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/namespace_test.go:23]: Maybe comment this a bit more? E.g., "we'll use the forest to create only full namespaces, and then manually turn them into subnamespaces by adding the annotations" or something like that?

File: incubator/hnc/internal/validators/namespace_test.go [hunk 6]
```diff
@@ -12,29 +12,53 @@ import (
...
 
-	t.Run("Delete namespace with owner annotation", func(t *testing.T) {
-		g := NewGomegaWithT(t)
-		req := &nsRequest{
-			ns: sub,
-			op: v1beta1.Delete,
-		}
+			// Construct the forest
+			f := foresttest.Create(tc.forest)
+			// Add anchor a to namespace c if it exists.
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/namespace_test.go:48]: This needs more comments somewhere, either here or maybe on lines 32 and 34. I can see that we _are_ adding the anchor if `c` exists, but _why_ are we doing this? (I figured it out, I'd just like it to be a bit better documented in the code).

**Comment by rjbez17** [comment]
lgtm after @adrianludwin comments are resolved.

File: incubator/hnc/internal/validators/namespace_test.go (hunk 4 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/namespace_test.go:26]: No, since parent and annotation always match.  There is a mistake here in the comment in line 34, the test case is subnamespaceOf b, but I commented "subnamespaceOf c". The comment is wrong. I updated it to b in the comment.

File: incubator/hnc/internal/validators/namespace_test.go (hunk 6 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/namespace_test.go]: I added the reason here in the same line.

File: incubator/hnc/internal/validators/namespace_test.go (hunk 4 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/namespace_test.go:25]: oh sorry somehow read it in the other way. will update..

**Comment by yiqigao217** [comment]
/retest

Similar to another flaky test #871 . I added the failed test logs to the issue.
